File: python_version/mixed_recall/linked_list_cycle/linked_list_cycle.py

# linked list cycle
# Write a function, linked_list_cycle, that takes
#  in the head of a linked list as an argument.
#   The function should return a boolean indicating whether
#    or not the linked list contains a cycle.


class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

# n = number of nodes
# Time: O(n)
# Space: O(1)


def linked_list_cycle_dual_pointers(head):


    return


# n = number of nodes
# Time: O(n)
# Space: O(n)
def linked_list_cycle_set(head):

    return 


a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')

a.next = b
b.next = c
c.next = d
d.next = b  # cycle

#         _______
#       /        \
# a -> b -> c -> d

print(linked_list_cycle_set(a))  # True
print(linked_list_cycle_dual_pointers(a))  # True

q = Node('q')
r = Node('r')
s = Node('s')
t = Node('t')
u = Node('u')

q.next = r
r.next = s
s.next = t
t.next = u
u.next = q  # cycle

#    ________________
#  /                 \
# q -> r -> s -> t -> u

print(linked_list_cycle_set(q))  # True
print(linked_list_cycle_dual_pointers(q))  # True

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d

print(linked_list_cycle_set(a))  # False
print(linked_list_cycle_dual_pointers(a))  # False

q = Node('q')
r = Node('r')
s = Node('s')
t = Node('t')
u = Node('u')

q.next = r
r.next = s
s.next = t
t.next = u
u.next = t  # cycle

#                   __
#                 /   \
# q -> r -> s -> t -> u

print(linked_list_cycle_set(q))  # True
print(linked_list_cycle_dual_pointers(q))  # True

p = Node('p')

# p

print(linked_list_cycle_set(p))  # False
print(linked_list_cycle_dual_pointers(p))  # False

print(linked_list_cycle_set(None))  # False
print(linked_list_cycle_dual_pointers(None))  # False

File: javascript_version/mixed_recall/linked_list_cycle/linked_list_cycle.js
// linked list cycle
// Write a function, linked_list_cycle, that takes
//  in the head of a linked list as an argument.
//   The function should return a boolean indicating whether
//    or not the linked list contains a cycle.

class Node {
    constructor (val) {
        this.val = val;
        this.next = null;
    }
}
// # n = number of nodes
// # Time: O(n)
// # Space: O(n)
const linkedListCycleSet = (head) => {
    // todo
    if (head === null) { return false; }
    let set = new Set();
    let curr = head;
    while (curr !== null) {
        if (set.has(curr)) { return true; }
        set.add(curr);
        curr = curr.next;
    }
    return false;
};




// # n = number of nodes
// # Time: O(n)
// # Space: O(1)

const linkedListCycleDualPointers = (head) => {
    // todo
    if (head === null) { return false; }
    let slow = head;
    let fast = head;
    let firstLoopBack = true;
    while (fast !== null && fast.next !== null) {
        if (fast === slow && firstLoopBack === false) { return true; }
        slow = slow.next;
        fast = fast.next.next;
        firstLoopBack = false;
    }
    return false;
};

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');

a.next = b;
b.next = c;
c.next = d;
d.next = b; // cycle

//         _______
//       /        \
// a -> b -> c -> d 

console.log(linkedListCycleSet(a));  // true
console.log(linkedListCycleDualPointers(a));  // true

const q = new Node('q');
const r = new Node('r');
const s = new Node('s');
const t = new Node('t');
const u = new Node('u');

q.next = r;
r.next = s;
s.next = t;
t.next = u;
u.next = q; // cycle

//    ________________
//  /                 \
// q -> r -> s -> t -> u 

console.log(linkedListCycleSet(q));  // true
console.log(linkedListCycleDualPointers(q));  // true

// const a = new Node('a');
// const b = new Node('b');
// const c = new Node('c');
// const d = new Node('d');

a.next = b;
b.next = c;
c.next = d;
d.next = null;
// a -> b -> c -> d 

console.log(linkedListCycleSet(a));  // false
console.log(linkedListCycleDualPointers(a));  // false

// const q = new Node('q');
// const r = new Node('r');
// const s = new Node('s');
// const t = new Node('t');
// const u = new Node('u');




q.next = r;
r.next = s;
s.next = t;
t.next = u;
u.next = t; // cycle

//                   __
//                 /   \
// q -> r -> s -> t -> u 

console.log(linkedListCycleSet(q));  // true
console.log(linkedListCycleDualPointers(q));  // true

const p = new Node('p');

// p

console.log(linkedListCycleSet(p)); // false
console.log(linkedListCycleDualPointers(p)); // false

console.log(linkedListCycleSet(null)); // false
console.log(linkedListCycleDualPointers(null)); // false

